### PR TITLE
Chronos:  Add TF2 support for `AutoLSTM` and `AutoS2S`

### PR DESCRIPTION
**Pull request by liangs6212**
`AutoLSTM` and `AutoSeq2Seq` support "keras" as backend, Model switching can be done by simply changing the backend.

- [x] : add model support 
- [x] : parameters description

### CODE REVIEW

**Comment by liangs6212** [comment]
@TheaperDeng `AutoModel` backend should  be "keras" or "tf2"?

File: python/chronos/src/bigdl/chronos/autots/model/auto_seq2seq.py [hunk 1]
```diff
@@ -79,10 +79,7 @@ def __init__(self,
             cluster, it defaults to "hdfs:///tmp/{name}".
         """
         super().__init__()
-        # todo: support search for past_seq_len.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/model/auto_seq2seq.py:82]: maybe we can leave this todo here.

File: python/orca/src/bigdl/orca/automl/model/base_keras_model.py [hunk 2]
```diff
@@ -24,17 +24,20 @@
 
 def check_tf_version():
     tf_version = tf.__version__
-    if tf_version >= "2.0":
-        raise RuntimeError(f"Currently running TensorFlow version {tf_version}. We only support"
-                           f"TensorFlow 1.x for now and has been tested on 1.15")
+    if tf_version <= "2.0":
+        raise RuntimeError(f"Currently running TensorFlow version {tf_version}. "
+                           "We only support TensorFlow 2.x for now.")
```
Comment (TheaperDeng) [inline python/orca/src/bigdl/orca/automl/model/base_keras_model.py:29]: I don't think we need to change this function since it is only called under `predict_with_uncertainty`, which is a function we don use.

File: python/chronos/src/bigdl/chronos/autots/model/auto_lstm.py [hunk 3]
```diff
@@ -90,10 +88,19 @@ def __init__(self,
...
+        if backend.startswith("torch"):
+            from bigdl.chronos.model.VanillaLSTM_pytorch import model_creator
+            model_builder = PytorchModelBuilder(model_creator=model_creator,
+                                                optimizer_creator=optimizer,
+                                                loss_creator=loss)
+        elif backend.startswith("keras"):
+            from bigdl.chronos.model.tf2.VanillaLSTM_keras import model_creator
+            model_builder = KerasModelBuilder(model_creator=model_creator,
+                                              optimizer=optimizer,
+                                              loss=loss)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/model/auto_lstm.py:101]: This automodel is inherited from BasePytorchAutomodel, may be we can create 2 automodels here called `AutoPytorchLSTM` and `AutoKerasLSTM`. `backend` parameters can be implemented under `AutoModelFactory`

File: python/orca/src/bigdl/orca/automl/model/base_keras_model.py [hunk 4]
```diff
@@ -224,13 +231,23 @@ def _get_required_parameters(self):
...
+        past_seq_len = self.config['past_seq_len']
+        input_feature_num = self.config['input_feature_num']
+        batch_size = self.config.get('batch_size', 32)
+        input_shape = tf.random.normal((batch_size, past_seq_len, input_feature_num))
+        return input_shape
 
 class KerasModelBuilder(ModelBuilder):
 
-    def __init__(self, model_creator):
+    def __init__(self, model_creator, optimizer, loss):
```
Comment (TheaperDeng) [inline python/orca/src/bigdl/orca/automl/model/base_keras_model.py:243]: why we need this `optimizer` and `loss`. I think there should be a good reason since may code change is related to this

File: python/orca/src/bigdl/orca/automl/model/base_keras_model.py (hunk 2 above)
Comment (liangs6212) [reply python/orca/src/bigdl/orca/automl/model/base_keras_model.py:29]: Ok, rollback

File: python/orca/src/bigdl/orca/automl/model/base_keras_model.py (hunk 4 above)
Comment (liangs6212) [reply python/orca/src/bigdl/orca/automl/model/base_keras_model.py:243]: `Modulebuilder` will init a model, if optimizer and loss are not specified, then the default parameters in `model_creator` will be used.

File: python/orca/src/bigdl/orca/automl/model/base_keras_model.py [hunk 5]
```diff
@@ -202,6 +208,9 @@ def state_dict(self):
     def load_state_dict(self, state):
         self.config = state["config"]
         self.model = self.model_creator(self.config)
+        if self.model.layers[0].weights.__len__() == 0:
+            input_shape = self._get_input_shape()
+            _, = self.model(input_shape)
```
Comment (TheaperDeng) [inline python/orca/src/bigdl/orca/automl/model/base_keras_model.py:213]: leave this bug to following PR. We may raise an issue first.

File: python/orca/src/bigdl/orca/automl/model/base_keras_model.py (hunk 4 above)
Comment (TheaperDeng) [reply python/orca/src/bigdl/orca/automl/model/base_keras_model.py:246]: optimizer and loss can be moved to search_space in automodels

File: python/chronos/src/bigdl/chronos/autots/model/base_automodel.py [hunk 6]
```diff
@@ -109,7 +105,7 @@ def predict_with_onnx(self, data, batch_size=32, dirname=None):
 
         Be sure to install onnx and onnxruntime to enable this function. The method
         will give exactly the same result as .predict() but with higher throughput
-        and lower latency.
+        and lower latency. keras will support onnx as soon as possible.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/model/base_automodel.py:108]: later.

File: python/chronos/src/bigdl/chronos/autots/model/base_automodel.py [hunk 7]
```diff
@@ -318,3 +316,30 @@ def get_best_config(self):
...
         return self.best_model
+
+    @staticmethod
+    def _dynamic_binding(obj, model_creator, optimizer, loss):
+        """
+        dynamic binding keras and pytorch method
+        :params obj: a AutoModel instance
+        :params model_creator: torch or kerasa model
+        :params optimizer: torch or keras optim
+        :prarms loss: torch or keras loss
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/model/base_automodel.py:327]: :return:

File: python/chronos/src/bigdl/chronos/autots/model/base_automodel.py [hunk 8]
```diff
@@ -318,3 +316,30 @@ def get_best_config(self):
...
+            model_builder = PytorchModelBuilder(model_creator=model_creator,
+                                                optimizer_creator=optimizer,
+                                                loss_creator=loss)
+            obj._DEFAULT_BEST_MODEL_DIR = "best_model.ckpt"
+            obj._DEFAULT_BEST_CONFIG_DIR = "best_config.json"
+        elif obj.backend.startswith("keras"):
+            from bigdl.orca.automl.model.base_keras_model import KerasModelBuilder
+            model_builder = KerasModelBuilder(model_creator=model_creator,
+                                              optimizer=optimizer,
+                                              loss=loss)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/model/base_automodel.py:340]: remove optimizer and loss

File: python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py [hunk 9]
```diff
@@ -17,76 +17,126 @@
...
+                 lstm_hidden_dim=128, lstm_layer_num=2, dropout=0.2):
+        self.input_feature_num = input_feature_num
+        self.lstm_hidden_dim = lstm_hidden_dim
+        self.lstm_layer_num = lstm_layer_num
+        self.dropout = dropout
+        self.encoder_lstm = []
+        for i in range(lstm_layer_num):
+            self.encoder_lstm.append(LSTM(self.lstm_hidden_dim,
+                                          return_sequences=True,
+                                          input_shape=(None, input_feature_num),
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py:34]: this input_feature_num should be dynamic (i.e. the value for this parameter is different for 1st layer and 2nd layer

File: python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py [hunk 10]
```diff
@@ -17,76 +17,126 @@
...
+                                          return_state=True,
+                                          dropout=dropout,
+                                          name="cus_encoder"+str(i+1)))
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py:37]: lstm_encoder

File: python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py [hunk 11]
```diff
@@ -17,76 +17,126 @@
...
+        self.output_feature_num = output_feature_num
+        self.lstm_hidden_dim = lstm_hidden_dim
+        self.lstm_layer_num = lstm_layer_num
+        self.dropout = dropout
+        self.decoder_lstm = []
+        for i in range(lstm_layer_num):
+            self.decoder_lstm.append(LSTM(self.lstm_hidden_dim,
+                                          return_sequences=True,
+                                          return_state=True,
+                                          input_shape=(1, output_feature_num),
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py:70]: should be lstm_hidden_dim

File: python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py [hunk 12]
```diff
@@ -17,76 +17,126 @@
...
+                                          dropout=dropout,
+                                          name="cus_decoder"+str(i+1)))
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/tf2/Seq2Seq_keras.py:72]: lstm_decoder

File: python/chronos/src/bigdl/chronos/model/tf2/VanillaLSTM_keras.py [hunk 13]
```diff
@@ -27,25 +27,26 @@ def __init__(self, input_dim, hidden_dim, layer_num, dropout, output_dim):
...
-        self.lstm_list.add(LSTM(self.hidden_dim[-1],
-                                dropout=dropout[-1],
-                                activation="linear",
-                                name="lstm_" + str(self.layer_num)))
-        self.fc = Dense(self.output_dim)
-        self.out_shape = Reshape((1, self.output_dim), input_shape=(self.output_dim,))
+            self.lstm_sequential.add(LSTM(self.hidden_dim[layer],
+                                          return_sequences=True,
+                                          dropout=self.dropout[layer],
+                                          input_shape=(1, self.input_dim),
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/tf2/VanillaLSTM_keras.py:35]: maybe we can delete this `input_shape`

File: python/chronos/src/bigdl/chronos/model/tf2/VanillaLSTM_keras.py [hunk 14]
```diff
@@ -27,25 +27,26 @@ def __init__(self, input_dim, hidden_dim, layer_num, dropout, output_dim):
...
+                                          name="lstm_" + str(layer+1)))
+        self.lstm_sequential.add(LSTM(self.hidden_dim[-1],
+                                      dropout=dropout[-1],
+                                      input_shape=(1, self.input_dim),
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/tf2/VanillaLSTM_keras.py:39]: maybe we can delete this `input_shape`

File: python/chronos/src/bigdl/chronos/model/tf2/VanillaLSTM_keras.py [hunk 15]
```diff
@@ -27,25 +27,26 @@ def __init__(self, input_dim, hidden_dim, layer_num, dropout, output_dim):
...
+        self.lstm_sequential.add(Reshape((1, self.output_dim), input_shape=(self.output_dim,)))
 
-    def call(self, input_seq):
-        lstm_out = input_seq
-        out = self.lstm_list(lstm_out)
-        out = self.fc(out)
-        out = self.out_shape(out)
+    def call(self, input_seq, training=False):
+        if not training:
+            # freeeze all lstm layers
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/model/tf2/VanillaLSTM_keras.py:46]: freeze all dropout in lstm layers

File: python/chronos/test/bigdl/chronos/autots/model/test_auto_lstm.py [hunk 16]
```diff
@@ -60,19 +60,21 @@ def valid_dataloader_creator(config):
                       shuffle=True)
 
 
-def get_auto_estimator():
+def get_auto_estimator(backend='torch'):
+    loss = torch.nn.MSELoss() if backend.startswith('torch') else "mse"
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/autots/model/test_auto_lstm.py:64]: `loss = "mse" if backend.startswith('keras') else torch.nn.MSELoss()`

File: python/orca/src/bigdl/orca/automl/model/base_keras_model.py (hunk 5 above)
Comment (liangs6212) [reply python/orca/src/bigdl/orca/automl/model/base_keras_model.py:213]: Ok, I temporarily removed the changes

File: python/chronos/src/bigdl/chronos/autots/model/base_automodel.py [hunk 17]
```diff
@@ -318,3 +316,33 @@ def get_best_config(self):
...
+            cls.search_space.update({"optimizer": cls.optimizer, "loss": cls.loss})
+            model_builder = KerasModelBuilder(model_creator=cls._model_creator)
+            cls._DEFAULT_BEST_MODEL_DIR = "best_keras_model.ckpt"
+            cls._DEFAULT_BEST_CONFIG_DIR = "best_keras_config.json"
+
+        from bigdl.orca.automl.auto_estimator import AutoEstimator
+        cls.auto_est = AutoEstimator(model_builder, **cls.auto_est_config)
+
+    cls.__init__ = __init__
+    return cls
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/model/base_automodel.py:348]: I am not quite sure why we need this decorator. It seems much easier if we can just move this to baseautomodel's `__init__`.

File: python/chronos/src/bigdl/chronos/autots/model/base_automodel.py [hunk 18]
```diff
@@ -18,13 +18,9 @@
...
 import json
 from bigdl.chronos.autots.utils import recalculate_n_sampling
-DEFAULT_BEST_MODEL_DIR = "best_model.ckpt"
-DEFAULT_BEST_CONFIG_DIR = "best_config.json"
 
 
-class BasePytorchAutomodel:
-    def __init__(self, **kwargs):
-        self.best_model = None
+class BaseAutomodelMix:
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/model/base_automodel.py:24]: it seems that many methods is only for pytorch, we may add
```python
if self.backend=="keras":
    # <raise error>
```

**Review by TheaperDeng — APPROVED** [review]
LGTM

**Review by TheaperDeng — APPROVED** [review]

**Review by TheaperDeng — APPROVED** [review]

**Comment by liangs6212** [comment]
http://10.112.231.51:18888/view/BigDL-2.0-PR/job/BigDL-Chronos-PR-Validation/758/
http://10.112.231.51:18889/view/BigDL-PR-Validation/job/BigDL-Chronos-PR-Validation/376/
http://10.112.231.51:18889/job/BigDL-PRVN-chronos-Python-Spark-3.1-py36-ray-part1/390/
